parser-typechecker/src/Unison/DeclPrinter.hs

```diff
@@ -133,55 +133,60 @@ prettyDataDecl (PrettyPrintEnvDecl unsuffixifiedPPE suffixifiedPPE) r name dd =
 --
 -- This function bails with `Nothing` if the names aren't an exact match for
 -- the expected record naming convention.
-fieldNames
-  :: forall v a . Var v
-  => PrettyPrintEnv
-  -> Reference
-  -> HashQualified Name
-  -> DataDeclaration v a
-  -> Maybe [HashQualified Name]
-fieldNames env r name dd = case DD.constructors dd of
-  [(_, typ)] -> let
-    vars :: [v]
-    vars = [ Var.freshenId (fromIntegral n) (Var.named "_") | n <- [0..Type.arity typ - 1]]
-    accessors :: [(v, Term.Term v ())]
-    accessors = DD.generateRecordAccessors (map (,()) vars) (HQ.toVar name) r
-    accessorsWithTypes :: [(v, Term.Term v (), Type.Type v ())]
-    accessorsWithTypes = accessors <&> \(v, trm) ->
-                      case Result.result (Typechecker.synthesize typecheckingEnv trm) of
-                        Nothing -> error $ "Failed to typecheck record field: " <> show v
-                        Just typ -> (v, trm, typ)
-    typeLookup :: TypeLookup v ()
-    typeLookup =
-      TypeLookup
-        { TypeLookup.typeOfTerms = mempty,
-          TypeLookup.dataDecls = Map.singleton r (void dd),
-          TypeLookup.effectDecls = mempty
-        }
-    typecheckingEnv :: Typechecker.Env v ()
-    typecheckingEnv =
-      Typechecker.Env
-        { Typechecker._ambientAbilities = mempty,
-          Typechecker._typeLookup = typeLookup,
-          Typechecker._termsByShortname = mempty
-        }
-    hashes = Hashing.hashTermComponents (Map.fromList . fmap (\(v, trm, typ) -> (v, (trm, typ))) $ accessorsWithTypes)
-    names = [ (r, HQ.toString . PPE.termName env . Referent.Ref $ DerivedId r)
-            | r <- (\(refId, _trm, _typ) -> refId) <$> Map.elems hashes ]
-    fieldNames = Map.fromList
-      [ (r, f) | (r, n) <- names
-               , typename <- pure (HQ.toString name)
-               , typename `isPrefixOf` n
-               , rest <- pure $ drop (length typename + 1) n
-               , (f, rest) <- pure $ span (/= '.') rest
-               , rest `elem` ["",".set",".modify"] ]
-    in if Map.size fieldNames == length names then
-         Just [ HQ.unsafeFromString name
-              | v <- vars
-              , Just (ref, _, _) <- [Map.lookup (Var.namespaced [HQ.toVar name, v]) hashes]
-              , Just name <- [Map.lookup ref fieldNames] ]
-       else Nothing
-  _ -> Nothing
+fieldNames ::
+  forall v a.
+  Var v =>
+  PrettyPrintEnv ->
+  Reference ->
+  HashQualified Name ->
+  DataDeclaration v a ->
+  Maybe [HashQualified Name]
+fieldNames env r name dd = do
```

Comment:
Most of the changes here are just re-formatting from a pure function into `Maybe` do-notation so I can easily fail the whole computation on a bad typecheck. I've noted the primary change with a comment.